http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/RuntimeServiceDbQueries.scala

```diff
@@ -307,33 +307,210 @@ object RuntimeServiceDbQueries {
+          String,
+          Boolean,
+          RuntimeConfigRecord,
+          RuntimeName,
+          RuntimeStatus,
+          Option[WorkspaceId],
+          Option[(String, String)]
+        )
+        records
+          .groupBy(_._1)
+          .map { case _ -> (values: Seq[ListRuntimesRecord]) =>
+            val labels: LabelMap = Map(values.mapFilter {
+              case (_, _, _, _, _, _, _, _, _, _, _, _, _, _, labelPair: Option[(String, String)]) => labelPair
+            }: _*)
+            val patchInProgress: Boolean = values.exists((value: ListRuntimesRecord) => value match {
+              case (_, _, _, _, _, _, _, _, _, inProgress: Boolean, _, _, _, _, _) => inProgress
+            })
+            values.head match {
+              case (
+                    id,
+                    cloudContextDb,
+                    cloudProvider,
+                    creatorEmail,
+                    createdDate,
+                    destroyedDate,
+                    dateAccessed,
+                    hostIp,
+                    internalId,
+                    _,
+                    runtimeConfigRecord,
+                    runtimeName,
+                    status,
+                    workspaceId,
+                    _
+                  ) =>
+                val auditInfo = AuditInfo(
+                  creatorEmail,
+                  createdDate,
+                  if (destroyedDate == dummyDate) None else Some(destroyedDate),
+                  dateAccessed
+                )
+                val cloudContext = (cloudProvider, cloudContextDb) match {
+                  case (CloudProvider.Gcp, CloudContextDb(value)) =>
+                    CloudContext.Gcp(GoogleProject(value))
+                  case (CloudProvider.Azure, CloudContextDb(value)) =>
+                    val context = AzureCloudContext.fromString(value).fold(s => throw new SQLDataException(s), identity)
+                    CloudContext.Azure(context)
+                }
+                val proxyUrl = Runtime.getProxyUrl(
+                  Config.proxyConfig.proxyUrlBase,
+                  cloudContext,
+                  runtimeName,
+                  Set.empty,
+                  hostIp,
+                  labels
+                )
+                val runtimeConfig = runtimeConfigRecord.runtimeConfig
+                val samResourceId = RuntimeSamResourceId(internalId)
+
+                ListRuntimeResponse2(
+                  id,
+                  auditInfo = auditInfo,
+                  cloudContext = cloudContext,
+                  clusterName = runtimeName,
+                  labels = labels,
+                  patchInProgress = patchInProgress,
+                  proxyUrl = proxyUrl,
+                  runtimeConfig = runtimeConfig,
+                  samResource = samResourceId,
+                  status = status,
+                  workspaceId = workspaceId
+                )
+            }
+          }
+          .toVector
+      }
```

Comment:
Is there a way to compare the query it prints out now with the query it used to print out? Mostly for curiosity as all of the unit tests are passing (and we had coverage for this code prior to the refactor)

Reply:
Will post shortly!

Reply:
Here is the new query: 

```
select 
  x2.x3,
  x2.x4,
  x2.x5,
  x2.x6,
  x2.x7,
  x2.x8,
  x2.x9,
  x2.x10,
  x2.x11,
  (
    case when (
      x12.`clusterId` is not null
    ) then 
      x12.`inProgress` 
    else 
      null 
    end
  ),
  x13.`id`,
  x13.`cloudService`,
  x13.`numberOfWorkers`,
  x13.`machineType`,
  x13.`diskSize`,
  x13.`bootDiskSize`,
  x13.`workerMachineType`,
  x13.`workerDiskSize`,
  x13.`numberOfWorkerLocalSSDs`,
  x13.`numberOfPreemptibleWorkers`,
  x13.`dataprocProperties`,
  x13.`persistentDiskId`,
  x13.`zone`,
  x13.`region`,
  x13.`gpuType`,
  x13.`numOfGpus`,
  x13.`componentGatewayEnabled`,
  x13.`workerPrivateAccess`,
  x13.`dateAccessed`,
  x2.x14,
  x2.x15,
  x2.x16,
  x17.`resourceId`,
  x17.`key`,
  x17.`value` 
from (
  (
    select 
      x18.x19 as x20,
      x18.x21 as x22,
      x18.x23 as x24,
      x18.x25 as x26,
      x18.x27 as x4,
      x18.x28 as x29,
      x18.x30 as x11,
      x18.x31 as x32,
      x18.x33 as x9,
      x18.x34 as x7,
      x18.x35 as x36,
      x18.x37 as x38,
      x18.x39 as x8,
      x18.x40 as x41,
      x18.x42 as x10,
      x18.x43 as x16,
      x18.x44 as x15,
      x18.x45 as x46,
      x18.x47 as x5,
      x18.x48 as x49,
      x18.x50 as x3,
      x18.x51 as x14,
      x18.x52 as x53,
      x18.x54 as x55,
      x18.x56 as x6,
      x18.x57 as x58 
    from (
      (
        select 
          x59.x60 as x19,
          x59.x61 as x21,
          x59.x62 as x23,
          x59.x63 as x25,
          x59.x64 as x27,
          x59.x65 as x28,
          x59.x66 as x30,
          x59.x67 as x31,
          x59.x68 as x33,
          x59.x69 as x34,
          x59.x70 as x35,
          x59.x71 as x37,
          x59.x72 as x39,
          x59.x73 as x40,
          x59.x74 as x42,
          x59.x75 as x43,
          x59.x76 as x44,
          x59.x77 as x45,
          x59.x78 as x47,
          x59.x79 as x48,
          x59.x80 as x50,
          x59.x81 as x51,
          x59.x82 as x52,
          x59.x83 as x54,
          x59.x84 as x56,
          x59.x85 as x57 
        from (
          (
            select 
              `startUserScriptUri` as x60,
              `stagingBucket` as x61,
              `defaultClientId` as x62,
              `customClusterEnvironmentVariables` as x63,
              `cloudContext` as x64,
              `welderEnabled` as x65,
              `internalId` as x66,
              `serviceAccount` as x67,
              `dateAccessed` as x68,
              `createdDate` as x69,
              `runtimeConfigId` as x70,
              `deletedFrom` as x71,
              `destroyedDate` as x72,
              `autopauseThreshold` as x73,
              `hostIp` as x74,
              `workspaceId` as x75,
              `status` as x76,
              `operationName` as x77,
              `cloudProvider` as x78,
              `userScriptUri` as x79,
              `id` as x80,
              `runtimeName` as x81,
              `kernelFoundBusyDate` as x82,
              `initBucket` as x83,
              `creator` as x84,
              `proxyHostName` as x85 
            from `CLUSTER` 
            where (
              `internalId` in (?, ?, ?, ?) and 
              (`workspaceId` in (?))
            ) union (
              select 
                `startUserScriptUri` as x60,
                `stagingBucket` as x61,
                `defaultClientId` as x62,
                `customClusterEnvironmentVariables` as x63,
                `cloudContext` as x64,
                `welderEnabled` as x65,
                `internalId` as x66,
                `serviceAccount` as x67,
                `dateAccessed` as x68,
                `createdDate` as x69,
                `runtimeConfigId` as x70,
                `deletedFrom` as x71,
                `destroyedDate` as x72,
                `autopauseThreshold` as x73,
                `hostIp` as x74,
                `workspaceId` as x75,
                `status` as x76,
                `operationName` as x77,
                `cloudProvider` as x78,
                `userScriptUri` as x79,
                `id` as x80,
                `runtimeName` as x81,
                `kernelFoundBusyDate` as x82,
                `initBucket` as x83,
                `creator` as x84,
                `proxyHostName` as x85 
              from `CLUSTER` 
              where false
            )
          ) 
        x59
      ) union (
        select 
          `startUserScriptUri` as x19,
          `stagingBucket` as x21,
          `defaultClientId` as x23,
          `customClusterEnvironmentVariables` as x25,
          `cloudContext` as x27,
          `welderEnabled` as x28,
          `internalId` as x30,
          `serviceAccount` as x31,
          `dateAccessed` as x33,
          `createdDate` as x34,
          `runtimeConfigId` as x35,
          `deletedFrom` as x37,
          `destroyedDate` as x39,
          `autopauseThreshold` as x40,
          `hostIp` as x42,
          `workspaceId` as x43,
          `status` as x44,
          `operationName` as x45,
          `cloudProvider` as x47,
          `userScriptUri` as x48,
          `id` as x50,
          `runtimeName` as x51,
          `kernelFoundBusyDate` as x52,
          `initBucket` as x54,
          `creator` as x56,
          `proxyHostName` as x57 
        from `CLUSTER` 
        where (
          (
            `internalId` in (
              ?, 
              ?, 
              ?, 
              ?
            )
          ) and (
            `cloudProvider` = 'GCP'
          )
        ) and (
          `cloudContext` in (
            ?
          )
        )
      )
    ) x18 
    where not (
      x18.x44 = 'Deleted'
    )
  ) union (
    select 
      `startUserScriptUri` as x20,
      `stagingBucket` as x22,
      `defaultClientId` as x24,
      `customClusterEnvironmentVariables` as x26,
      `cloudContext` as x4,
      `welderEnabled` as x29,
      `internalId` as x11,
      `serviceAccount` as x32,
      `dateAccessed` as x9,
      `createdDate` as x7,
      `runtimeConfigId` as x36,
      `deletedFrom` as x38,
      `destroyedDate` as x8,
      `autopauseThreshold` as x41,
      `hostIp` as x10,
      `workspaceId` as x16,
      `status` as x15,
      `operationName` as x46,
      `cloudProvider` as x5,
      `userScriptUri` as x49,
      `id` as x3,
      `runtimeName` as x14,
      `kernelFoundBusyDate` as x53,
      `initBucket` as x55,
      `creator` as x6,
      `proxyHostName` as x58 
    from `CLUSTER` 
    where (
      (
        `cloudProvider` = 'GCP'
      ) and false
    ) and (
      not (
        `status` = 'Deleted'
      )
    )
  )
) x2 
inner join `RUNTIME_CONFIG` x13 on 
  x2.x36 = x13.`id` 
left outer join `CLUSTER_PATCH` x12 on 
  x2.x3 = x12.`clusterId` 
left outer join `LABEL` x17 on (
  x2.x3 = x17.`resourceId`
) and (
  x17.`resourceType` = 'runtime'
)
```

Reply:
And here is another new query from a different test run:

```
select x2.x3,
  x2.x4,
  x2.x5,
  x2.x6,
  x2.x7,
  x2.x8,
  x2.x9,
  x2.x10,
  x2.x11,
  (
 case when (
 x12.`clusterId` is not null
) then x12.`inProgress` else null end
),
  x13.`id`,
  x13.`cloudService`,
  x13.`numberOfWorkers`,
  x13.`machineType`,
  x13.`diskSize`,
  x13.`bootDiskSize`,
  x13.`workerMachineType`,
  x13.`workerDiskSize`,
  x13.`numberOfWorkerLocalSSDs`,
  x13.`numberOfPreemptibleWorkers`,
  x13.`dataprocProperties`,
  x13.`persistentDiskId`,
  x13.`zone`,
  x13.`region`,
  x13.`gpuType`,
  x13.`numOfGpus`,
  x13.`componentGatewayEnabled`,
  x13.`workerPrivateAccess`,
  x13.`dateAccessed`,
  x2.x14,
  x2.x15,
  x2.x16,
  x17.`resourceId`,
  x17.`key`,
  x17.`value` from (
 (
 select x18.x19 as x20,
  x18.x21 as x22,
  x18.x23 as x24,
  x18.x25 as x26,
  x18.x27 as x4,
  x18.x28 as x29,
  x18.x30 as x11,
  x18.x31 as x32,
  x18.x33 as x9,
  x18.x34 as x7,
  x18.x35 as x36,
  x18.x37 as x38,
  x18.x39 as x8,
  x18.x40 as x41,
  x18.x42 as x10,
  x18.x43 as x16,
  x18.x44 as x15,
  x18.x45 as x46,
  x18.x47 as x5,
  x18.x48 as x49,
  x18.x50 as x3,
  x18.x51 as x14,
  x18.x52 as x53,
  x18.x54 as x55,
  x18.x56 as x6,
  x18.x57 as x58 from (
 (
 select x59.x60 as x19,
  x59.x61 as x21,
  x59.x62 as x23,
  x59.x63 as x25,
  x59.x64 as x27,
  x59.x65 as x28,
  x59.x66 as x30,
  x59.x67 as x31,
  x59.x68 as x33,
  x59.x69 as x34,
  x59.x70 as x35,
  x59.x71 as x37,
  x59.x72 as x39,
  x59.x73 as x40,
  x59.x74 as x42,
  x59.x75 as x43,
  x59.x76 as x44,
  x59.x77 as x45,
  x59.x78 as x47,
  x59.x79 as x48,
  x59.x80 as x50,
  x59.x81 as x51,
  x59.x82 as x52,
  x59.x83 as x54,
  x59.x84 as x56,
  x59.x85 as x57 from (
 (
 select `startUserScriptUri` as x60,
  `stagingBucket` as x61,
  `defaultClientId` as x62,
  `customClusterEnvironmentVariables` as x63,
  `cloudContext` as x64,
  `welderEnabled` as x65,
  `internalId` as x66,
  `serviceAccount` as x67,
  `dateAccessed` as x68,
  `createdDate` as x69,
  `runtimeConfigId` as x70,
  `deletedFrom` as x71,
  `destroyedDate` as x72,
  `autopauseThreshold` as x73,
  `hostIp` as x74,
  `workspaceId` as x75,
  `status` as x76,
  `operationName` as x77,
  `cloudProvider` as x78,
  `userScriptUri` as x79,
  `id` as x80,
  `runtimeName` as x81,
  `kernelFoundBusyDate` as x82,
  `initBucket` as x83,
  `creator` as x84,
  `proxyHostName` as x85 from `CLUSTER` where (
 `internalId` in (
 ?,
  ?
)
) and (
 `workspaceId` in (
 ?
)
)
) union (
 select `startUserScriptUri` as x60,
  `stagingBucket` as x61,
  `defaultClientId` as x62,
  `customClusterEnvironmentVariables` as x63,
  `cloudContext` as x64,
  `welderEnabled` as x65,
  `internalId` as x66,
  `serviceAccount` as x67,
  `dateAccessed` as x68,
  `createdDate` as x69,
  `runtimeConfigId` as x70,
  `deletedFrom` as x71,
  `destroyedDate` as x72,
  `autopauseThreshold` as x73,
  `hostIp` as x74,
  `workspaceId` as x75,
  `status` as x76,
  `operationName` as x77,
  `cloudProvider` as x78,
  `userScriptUri` as x79,
  `id` as x80,
  `runtimeName` as x81,
  `kernelFoundBusyDate` as x82,
  `initBucket` as x83,
  `creator` as x84,
  `proxyHostName` as x85 from `CLUSTER` where false
)
) x59
) union (
 select `startUserScriptUri` as x19,
  `stagingBucket` as x21,
  `defaultClientId` as x23,
  `customClusterEnvironmentVariables` as x25,
  `cloudContext` as x27,
  `welderEnabled` as x28,
  `internalId` as x30,
  `serviceAccount` as x31,
  `dateAccessed` as x33,
  `createdDate` as x34,
  `runtimeConfigId` as x35,
  `deletedFrom` as x37,
  `destroyedDate` as x39,
  `autopauseThreshold` as x40,
  `hostIp` as x42,
  `workspaceId` as x43,
  `status` as x44,
  `operationName` as x45,
  `cloudProvider` as x47,
  `userScriptUri` as x48,
  `id` as x50,
  `runtimeName` as x51,
  `kernelFoundBusyDate` as x52,
  `initBucket` as x54,
  `creator` as x56,
  `proxyHostName` as x57 from `CLUSTER` where (
 (
 `internalId` in (
 ?,
  ?
)
) and (
 `cloudProvider` = 'GCP'
)
) and (
 `cloudContext` in (
 ?
)
)
)
) x18 where not (
 x18.x44 = 'Deleted'
)
) union (
 select `startUserScriptUri` as x20,
  `stagingBucket` as x22,
  `defaultClientId` as x24,
  `customClusterEnvironmentVariables` as x26,
  `cloudContext` as x4,
  `welderEnabled` as x29,
  `internalId` as x11,
  `serviceAccount` as x32,
  `dateAccessed` as x9,
  `createdDate` as x7,
  `runtimeConfigId` as x36,
  `deletedFrom` as x38,
  `destroyedDate` as x8,
  `autopauseThreshold` as x41,
  `hostIp` as x10,
  `workspaceId` as x16,
  `status` as x15,
  `operationName` as x46,
  `cloudProvider` as x5,
  `userScriptUri` as x49,
  `id` as x3,
  `runtimeName` as x14,
  `kernelFoundBusyDate` as x53,
  `initBucket` as x55,
  `creator` as x6,
  `proxyHostName` as x58 from `CLUSTER` where (
 (
 `cloudProvider` = 'GCP'
) and false
) and (
 not (
 `status` = 'Deleted'
)
)
)
) x2 inner join `RUNTIME_CONFIG` x13 on x2.x36 = x13.`id` left outer join `CLUSTER_PATCH` x12 on x2.x3 = x12.`clusterId` left outer join `LABEL` x17 on (
 x2.x3 = x17.`resourceId`
) and (
 x17.`resourceType` = 'runtime'
)
```

Reply:
thanks! and ... wow. just wow. 🤪